wikimedia/wikipedia-ios
# Determine new notifications in extension and display

**Phabricator:** 
https://phabricator.wikimedia.org/T288773
https://phabricator.wikimedia.org/T285353
& Some of https://phabricator.wikimedia.org/T287033

### Notes
This extends the work done in #4026 to pull the latest unread push notifications from the notifications API, filters out previously seen notifications via an app container cache file and displays the results. If the result is a single new unread notification, it modifies the push content to display it's `header` value. There is also some extra handling for talk page notification types (adds subtitle, bundles if necessary) based on the [Figma mocks](https://www.figma.com/file/cedgOU5CyOR0UVqtjDOvzE/iOS-Notifications?node-id=549%3A1752). For all other situations, it defaults to "New activity on Wikipedia" text.

After review, please wait for #4044 to be reviewed and merged. Then update this with feature/notifications and merge into that feature branch.

These steps should be tested through the [Beta Cluster](https://en.wikipedia.beta.wmflabs.org/wiki/Main_Page) via the Staging scheme. 

### Test Steps
1. Fresh install from this branch and log in from an account that has all checkboxes checked on desktop Wikipedia notifications preferences. Go to Settings > Notifications and enable notifications.
2. Do something to trigger a notification to yourself from another account on desktop Wikipedia. I tend to test by thanking an edit, undoing an edit, or writing on my talk page with another account.
3. Wait for notifications to appear on your phone. On Staging scheme / beta cluster desktop it takes about 10 seconds for me lately.
4. The first notification will likely say "New activity on Wikipedia", since all notifications appear new to the device, and it sees it as condensing multiple notification types.
5. Trigger another notification (thanks or undo edit).
6. Now you should see a push with content specific push notification content.
7. Go to your user talk page and write on it.
8. Now you should see a push with content specific push notification content, plus a "New message" subtitle.
9. Get a couple of browser tabs open with editing your talk page. Try to post two updates back to back to trigger a batched talk page notification
10. Now you should see a push with batched content, "N new messages on {Title}", as well a "New messages" subtitle.
11. Go to desktop notifications preferences, uncheck "Thanks" from the app column.
12. Trigger a thanks again to yourself. Confirm no notification appears on your device.
